Cargo, Crates, and Workspaces (#43)
ferrous-systems/rust-training: training-slides/src/using-cargo.md

* `cargo clippy` - runs a linter
* `cargo test` - builds your project if necessary and runs tests
    * by default runs unit tests, integration tests, and documentation tests
    * you can select which tests to run
* `cargo build --release` - produces an optimized version of your application or library

## Cargo commands (cont)

* There are many more!
    * `cargo bench` - builds an optimized version of your project and runs benchmarks
    * `cargo doc --open` - builds documentation for your project *and all its dependencies* and opens it in a browser
    * `cargo run --example ...` - runs an example from your `examples/` directory

See [Cargo Book](https://doc.rust-lang.org/cargo/commands/index.html) for more.

## Cargo command arguments

Most cargo commands accept a few common arguments:

* `+toolchain`
* `--target`
* `--features`, `--all-features`, and `--no-default-features`
* `--timings`

## Putting it all together:

`cargo +nightly run --target x86_64-apple-darwin --features "a b c dependency/feature" --timings`

* use nightly Rust
* enable features `a`, `b`, `c`, and a `feature` feature of a `dependency` crate

> * (assuming we use Apple Silicon computer) build a macOS executable for x86 processor and run it using build-in emulation (Rosetta2)

Reviewer: ```suggestion
* (assuming we use an Apple Silicon computer) build a macOS executable for x86 processor and run it using built-in emulation (Rosetta2)
```

[dependencies]
thiserror = { workspace = true }
service_a = { path = "../service_a" }
...
```

## Cargo commands for workspaces

* `cargo run --bin main_app`
* `cargo test -p service_a`

## Understanding Rust project build time

* Cargo keeps track of changes you make and only rebuilds what is necessary
* When building a crate `rustc` can do most of work in parallel, but some steps still require synchronization
* Depending on type of a build times spent in different build phases may be vastly different.
    * debug vs release
    * various flags for `rustc` and LLVM
    * a build from scratch vs an incremental build

## Producing a build timings report

`rm -rf target/debug && cargo build --timings`

```text
.
└── target/
    ├── cargo-timings/
    │   ├── cargo-timings.html
    │   └── cargo-timings-<timestamp>.html
    ├── debug/
    └── ...
```

## Timings Report

![Cargo Build Report for Rust Analyzer](../book/images/rust-analyzer-cargo-build-timings.png)

## Reading the report

* Cargo can't start building a crate until all its dependencies have been built.
    * Cargo only waits for `rustc` to produce an LLVM IR, further compilation by LLVM can run in background (purple)

> * A crate can't start building until its `build.rs` is build and finish running (yellow)

Reviewer 2: ```suggestion
* A crate can't start building until its `build.rs` is built and finishes running (yellow)
```

# Cargo Dependencies and Workspaces

# Crates

> * Rust libraries are called `crates`

Reviewer 2: I don't know if I like the word library here, but I can't think of another word for "a lump of code".

Fun fact: IMHO crates.io is not a crate repository, it is a package repository - a package being the thing defined by the `Cargo.toml` file, which may contain crates.

Reviewer 2: I think maybe:

* Rust code is arranged into packages
* A package can contain a single *library crate*, and/or 1 or more *binary crates*
* The package is described with a file called `Cargo.toml`
* The crates in that package are built using `cargo`

Author: I went through all mentions of crates and packages and reworked all wording that could be lousy with the terms.

* Crates are built using `cargo`

> * A crate can produce a single library and / or 1 or more executables

Reviewer 2: *Technically*, a *package* can contain a single *library crate* and/or 1 more more *binary (i.e. executable) crates*

* A single crate can contain multiple modules

> * Multiple crates can be combined into a single Cargo workspace

Reviewer 2: ```suggestion
* Multiple packages can be combined into a single Cargo workspace
```

# Cargo Dependencies and Workspaces

# Crates and Packages

* Rust code is arranged into packages
* A package is described by a `Cargo.toml` file
* Building a package can produce a single library, and 0 or more executables

>     * these build artifacts are called "crates"

Reviewer 2: ```suggestion
    * these are called *crates*
```
